### DISCUSSION TITLE
Server-sent events (SSE)

### DISCUSSION THREAD

**Discussion by ghost** [adonisjs/core#1311] [Ideas]
Would it be possible to add support for server-sent events? They would be very useful for notifying changes for example.

**Comment by popovicieduard** [comment]
There is support for websockets if that is what you are looking for: https://adonisjs.com/docs/4.1/websocket

**Comment by thetutlage** [comment]
@qjnr Have you tried adding them manually?

**Reply by ghost** [reply]
Yes, I've tried adding them but I could not get it to work. I've tried to implement a simple countdown as a test.

Here is the code from my `start/routes.ts` file:
```ts
import Route from '@ioc:Adonis/Core/Route'
import { ResponseContract } from '@ioc:Adonis/Core/Response'

Route.on('/').render('index')

function countdown(response: ResponseContract, count: number) {
  console.log(`Countdown: ${count}`)
  response.send(`data: ${count}\n\n`)
  if (count) {
    return new Promise((resolve) => {
      setTimeout(() => resolve(countdown(response, count - 1)), 1000)
    })
  }
}

Route.get('/countdown', async ({ response }) => {
  response.header('content-type', 'text/event-stream')
  response.header('cache-control', 'no-cache')
  response.header('connection', 'keep-alive')
  //response.flushHeaders(200)

  await countdown(response, 10)
})
```
And here is the code for my `resources/views/index.edge` file:
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Adonis SSE Test</title>
</head>
<body>
  <ul id="data"></ul>

  <script>
    const eventSource = new EventSource('/countdown')
    const dataList = document.querySelector('#data')

    eventSource.addEventListener('open', (event) => {
      console.log(event)
    })
    eventSource.addEventListener('message', (event) => {
      console.log(event)
      dataList.insertAdjacentHTML('beforeend', `
        <li>${event.data}</li>
      `)
    })
  </script>
</body>
</html>
```

What's happening is that the `onopen` event (client-side) only fires after the countdown has finished, and thus the only message is the last number (which is zero). I think I need to flush the response or something everytime it counts down, because currently, it seems to only send the response after the request handler is finished, which is after the promise resolves.

If I don't use promises, only the first number gets sent and some weird stuff is happening on the server-side. I'm using a `console.log` statement to view the current state of the counter, and it shows that it's calling lots of different countdown at the same time. I believe it's because the client tries to reconnect automatically after the countdown is finished, but the response has ended early so only the 10 gets sent. Then the client reconnects, and the server starts a new countdown, while the previous one wasn't finished already.

I hope my explanation of what's happening is clear. Is there any way to solve this problem (sending a response before the handler is finished?)
